### ISSUE TITLE
SO ECP problem

### ISSUE THREAD

**Issue by dlmv** [pyscf/pyscf#744]
Dear developers,

I tried to use PYSCF with spin-orbit ECP and it seems from energy comparison that it takes into account only the component with max L (ul) component (and applies it wrongly).
As an example, for Pb2+ ion I got that the total energy is only influenced by ul-component:
https://u.pcloud.link/publink/show?code=kZ7adpXZHOPfskzmJskCuh6egAnI3SwKPbpX

However, when the same calculations are performed with other software, such as DIRAC, the g-component of SO almost does not affect total energy, but components with lower L do, which can be easily explained by zero occupancy of g-orbitals.
I also got similar results for some more structures both for HF and DFT.

So, either I did some mistakes when preparing inputs, or there is a bug in reading or applying SO ECP.

**Comment by sunqm** [comment]
The RHF code only supports the scalar part of ECP. The SOC part of an ECP can be computed using the integral code `mol.intor('ECPso')` . This contribution needs to be added to a different method, like GHF.
```
mf = mol.GHF()
s =  lib.PauliMatrices * .5
ecpso = np.einsum('sxy,spq->pxqy', s, mol.intor('ECPso'))
hcore = mf.get_hcore() + ecpso.reshape(mol.nao*2,mol.nao*2)
mf.get_hcore = lambda *args: hcore
mf.kernel()
```

Relevant examples are https://github.com/pyscf/pyscf/blob/master/examples/gto/20-soc_ecp.py and https://github.com/pyscf/pyscf/blob/master/examples/scf/44-soc_ecp.py

**Comment by dlmv** [comment]
Thank you for explanation.
However, I tried new input on base of 44-soc_ecp.py and got positive energy for my system.
I also repeated the calculation with a smaller basis, and got very different results, so it is possible that the problem is related to basis size.
https://u.pcloud.link/publink/show?code=kZm2GpXZUG6W1Qf3NWR78O1OvfWJIfQ1jE5y

Is there a way to overcome this problem?

**Comment by bogdanoff** [comment]
It seems like your calculations do not converge. I repeated them with the Stuttgart SO ECP and also found problems with convergence, which can be solved by using the second-order SCF solver, for example by adding `.newton()` decorator to HF object.
I also find it useful to start with non-relativistic SCF calculation.

Stuttgart basis gives  E_trip - E_sing 7.61283 eV with RHF and 7.54908 eV with GHF
> (RHF S=0) converged SCF energy = -191.109572620205
> (GHF S=0) converged SCF energy = -191.109572620206  <S^2> = -8.8817842e-16  2S+1 = 1
> (RHF S=1) converged SCF energy = -190.82980643892
> (GHF S=1) converged SCF energy = -190.832149087218  <S^2> = 2.0030869  2S+1 = 3.0020572

Your basis gives 7.70096 eV and 7.63501 eV with RHF and GHF
> converged SCF energy = -191.330589316205
> converged SCF energy = -191.330589316278  <S^2> = -5.3290705e-15  2S+1 = 1
> converged SCF energy = -191.047584142622
> converged SCF energy = -191.0500077802  <S^2> = 2.0031789  2S+1 = 3.0021185

Here is my input:
```
from pyscf import gto, lib

mol = gto.M(
    verbose=5,
    atom='Pb    0 0 0',
    charge=2,
    basis={'Pb': "cc-pv5z-pp"},
    ecp={'Pb': gto.basis.parse_ecp('''                                                                                                                                                                             
ECP                                                                                                                                                                                                                
Pb nelec 60                                                                                                                                                                                                        
Pb ul                                                                                                                                                                                                              
2       1.0000000              0.0000000                                                                                                                                                                           
Pb S                                                                                                                                                                                                               
2      12.2963030            281.2854990                                                                                                                                                                           
2       8.6326340             62.5202170                                                                                                                                                                           
Pb P                                                                                                                                                                                                               
2      10.2417900             72.2768970  -144.553795                                                                                                                                                              
2       8.9241760            144.5910830  144.591083                                                                                                                                                               
2       6.5813420              4.7586930  -9.517385                                                                                                                                                                
2       6.2554030              9.9406210  9.940621                                                                                                                                                                 
Pb D                                                                                                                                                                                                               
2       7.7543360             35.8485070  -35.848507                                                                                                                                                               
2       7.7202810             53.7243420  35.816228                                                                                                                                                                
2       4.9702640             10.1152560  -10.115256                                                                                                                                                               
2       4.5637890             14.8337310  9.889154                                                                                                                                                                 
Pb F                                                                                                                                                                                                               
2       3.8875120             12.2098920  -8.139928                                                                                                                                                                
2       3.8119630             16.1902910  8.095145                                                                                                                                                                 
Pb G                                                                                                                                                                                                               
2       5.6915770             -9.0966650  4.548332                                                                                                                                                                 
2       5.7155670            -11.5319960  -4.612798                                                                                                                                                                
END                                                                                                                                                                                                                
    ''')},
    spin=0
)

mf1 = mol.RHF().newton()
mf1.chkfile = "stutt.chk"
mf1.kernel()

mf = mol.GHF().newton()
s = .5 * lib.PauliMatrices
ecpso = lib.einsum('sxy,spq->pxqy', s, mol.intor('ECPso'))
hcore = mf.get_hcore()
hcore = hcore + ecpso.reshape(hcore.shape)
mf.get_hcore = lambda *args: hcore
dm = mf.from_chk(mf1.chkfile)
mf.kernel(dm)


mol.spin = 2

mf1 = mol.RHF().newton()
mf1.chkfile = "stutt.chk"
mf1.kernel()

mf = mol.GHF().newton()
s = .5 * lib.PauliMatrices
ecpso = lib.einsum('sxy,spq->pxqy', s, mol.intor('ECPso'))
hcore = mf.get_hcore()
hcore = hcore + ecpso.reshape(hcore.shape)
mf.get_hcore = lambda *args: hcore
dm = mf.from_chk(mf1.chkfile)
mf.kernel(dm)
```

**Comment by dlmv** [comment]
> mf1 = mol.RHF().newton()
> mf1.chkfile = "stutt.chk"
> mf1.kernel()
> 
> mf = mol.GHF().newton()
> s = .5 * lib.PauliMatrices
> ecpso = lib.einsum('sxy,spq->pxqy', s, mol.intor('ECPso'))
> hcore = mf.get_hcore()
> hcore = hcore + ecpso.reshape(hcore.shape)
> mf.get_hcore = lambda *args: hcore
> dm = mf.from_chk(mf1.chkfile)
> mf.kernel(dm)

Thank you, with this additions I managed to achieve convergence even with my basis and ECP.
However, the initial problem persists: removing SO part from Stuttgart ECP does not change the total energy. You can try it by replacing the corresponding column with zeros.

**Comment by bogdanoff** [comment]
Isn't it related to the absence of first order SOC in the 1S singlet state? For the triplet state there is some stabilization with GHF.

**Comment by dlmv** [comment]
While there is no splitting in this state, there should be a correction to total energy. I am afraid I don't have enough theoretical knowledge to deduce it, but both programs that I have access to (one of them is DIRAC) show decrease of the total energy by about 0.2 AU when SO is included.
Additionally, there is also no energy difference for Pb+1 and Pb0 with and without SO.
However, when a fake SO projector for ul(H) is introduced, the total energy changes in all cases, which does not seem to be consistent with any assumption.

**Comment by sunqm** [comment]
The ECP SOC integrals are abnormally large using the SOC coefficients provided in your input.
There are different conventions on the coefficients. I guess there might be convention mismatch in this ecp dataset.

**Comment by dlmv** [comment]
I think it is not the case, as I checked this ECP in other software, managed to achieve reasonable convergence in PYSCF with @bogdanoff's tips, and, last but not least, the Stuttgart ECP yields the same incorrect dependence on SO components, so that the initial problem persists.

**Comment by xubwa** [comment]
But in dirac, it's always all electron calculation, maybe you should compare with NWChem's spin-orbit ecp results.

**Comment by dlmv** [comment]
>But in dirac, it's always all electron calculation, maybe you should compare with NWChem's spin-orbit ecp results.
http://www.diracprogram.org/doc/master/molecule_and_basis/molecule_with_ecp.html
http://www.diracprogram.org/doc/release-18/tutorials/ecp/first.html
?

**Comment by xubwa** [comment]
Sorry, I didn't know that before.

**Comment by sunqm** [comment]
> I think it is not the case, as I checked this ECP in other software, managed to achieve reasonable convergence in PYSCF with @bogdanoff's tips, and, last but not least, the Stuttgart ECP yields the same incorrect dependence on SO components, so that the initial problem persists.

Stuttgart ECP-SO parameters are not currently available in the program. If these parameters are obtained from different packages, the convention of ECP SOC coefficients might be still the source of the problem.

**Comment by sunqm** [comment]
Based on Dirac's document, SO parameters should be compatible. Pyscf ECP integral requires the factor 2/2l+1 being multiplied in the SO parameters

**Comment by xubwa** [comment]
I believe the pyscf soecp function can use the stuttgart ecp correctly since I have reproduced molpro's zero field splitting using pyscf's soecp integrals+shci.

**Comment by dlmv** [comment]
>Stuttgart ECP-SO parameters are not currently available in the program.

I took them from the message in this thread. If there is SO-ECP for Pb already prepared from PYSCF, I can repeat the calculations with it.

But, I don't think this is a source of problem, as the problem is not quantitative, but qualitative. It is not the absolute values but the overall dependence on SO parameters is incorrect. When two-component SO-ECP calculation is performed, the total energy should decrease compared to AREP-only calculation, and the effect should be largest for p-component and almost zero for f and g components. This is observed in DIRAC and other (non-public) program I used. However, in the PYSCF, only the SO-component with l=ul has influence on the total energy, which is exact opposite behavior.
